### ISSUE TITLE
New Template Format

### ISSUE THREAD

**Issue by lxt2** [cardigann/cardigann#380]
I've been away for a while, so catching up. Interesting to see how you've adopted the definition format in Jackett @kaso17. 

The more I think about it, the more I think it makes sense to offer some sort of pre-processing of pages with Javascript and to standardize on using cookies for login. That would let us keep the definition format simpler, which was the aim. As it stands, it's starting to get seriously complicated. 

Perhaps we could do this incrementally by adding a `script` block to the templates? The scripts could be used to preprocess the DOM/HTML that was fetched, with the option of doing `fetch`/xhr calls to add more data if needed. 

I'm currently looking at using https://github.com/deoxxa/ottoext to provide some Node.JS-like event loop and a `fetch` api. Paired with something like https://github.com/tmpvar/jsdom and a selector engine and most of the really complex query and transformation stuff could be moved to the script blocks.

**Comment by kaso17** [comment]
Doing cookie login only is a bad idea, plenty of trackers need a new login/cookies after a specific time and/or IP change.

Using JS as a preprocessor to generate a HTML page which is then parsed by the current parser again doesn't make sense. In that case I would just do the complete parsing in JS and just add some JS method to send releases, etc. back to cardigann. That would be the solution for complex trackers which we would implement as native C# indexers in Jackett.
Besides that I would use JS to replace the template engine and probably would also support it on the selector level (run javascript in the context of the selector result) and also as filter (text input/output).

In Jackett we use https://anglesharp.github.io/ for HTML DOM parsing/CSS selectors. It also has a javaScript extension (based on the jInt interpreter). That obviously would be the first choice. Doing dom parsing in JS like you suggested sounds slow. With the JavaScript/AngleSharp integration it's not needed (the JS has access to the DOM parsed by AngleSharp).

**Comment by lxt2** [comment]
Yup, think I agree. I've struggled to find a good format for templating with JS.
